File: 백준/Silver/11725. 트리의 부모 찾기/트리의 부모 찾기.py
import sys
sys.setrecursionlimit(10 ** 9)
input = sys.stdin.readline

def dfs(start, tree, parents):
    for i in tree[start]: #연결된 노드 모두 탐색
        if parents[i] == 0: #한번도 방문한 적이 없다면
            parents[i] = start #부모노드 저장
            dfs(i, tree, parents) #해당 노드의 자식노드 탐색

n = int(input())
tree = [[] for _ in range(n+1)]
for _ in range(n-1):
    a, b = map(int, input().split())
    tree[a].append(b)
    tree[b].append(a)

#부모 저장 리스트
parents = [0 for _ in range(n+1)]

dfs(1, tree, parents)

for i in range(2, n+1):
    print(parents[i])

File: 백준/Silver/11725. 트리의 부모 찾기/트리의 부모 찾기.java
import java.io.*;
import java.util.*;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));

        int N = Integer.parseInt(br.readLine());
        int[] parent = new int[N + 1];
        Map<Integer, List<Integer>> tree = new HashMap<>();

        for (int i = 0; i < N - 1; i++) {
            StringTokenizer st = new StringTokenizer(br.readLine());
            int a = Integer.parseInt(st.nextToken());
            int b = Integer.parseInt(st.nextToken());
            if (tree.containsKey(a)) {
                tree.get(a).add(b);
            } else {
                tree.put(a, new ArrayList<>(Arrays.asList(b)));
            }
            if (tree.containsKey(b)) {
                tree.get(b).add(a);
            } else {
                tree.put(b, new ArrayList<>(Arrays.asList(a)));
            }
        }
        dfs(1, tree, parent);
        for (int i = 2; i < parent.length; i++) {
            System.out.println(parent[i]);
        }

        bw.close();
        br.close();
    }

    public static void dfs(int start, Map<Integer, List<Integer>> tree, int[] parent) {
        for (Integer i : tree.get(start)) {
            if (parent[i] == 0) {
                parent[i] = start;
                dfs(i, tree, parent);
            }
        }
    }
}
